pkg/liveness/liveness.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package liveness
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"time"
+
+	"k8s.io/klog"
+
+	connlib "github.com/kubernetes-csi/csi-lib-utils/connection"
+	"github.com/kubernetes-csi/csi-lib-utils/rpc"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	liveness = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "csi",
+		Name:      "liveness",
+		Help:      "Liveness Probe",
+	})
+)
+
+func getLiveness(endpoint string) {
+
+	csiConn, err := connlib.Connect(endpoint)
+	if err != nil {
+		// connlib should retry forever so a returned error should mean
+		// the grpc client is misconfigured rather than an error on the network
+		klog.Infof("failed to establish connection to CSI driver: %v", err)
+	}
+
+	log.Printf("calling CSI driver to discover driver name")
+	csiDriverName, err := rpc.GetDriverName(context.Background(), csiConn)
+	if err != nil {
+		klog.Infof("failed to get CSI driver name: %v", err)
+	}
+	log.Printf("CSI driver name: %q", csiDriverName)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
+
+	log.Printf("Sending probe request to CSI driver %q", csiDriverName)
+	ready, err := rpc.Probe(ctx, csiConn)
+	if err != nil {
+		liveness.Set(0)
+		klog.Infof("health check failed: %v", err)
+		return
+	}
+
+	if !ready {
+		liveness.Set(0)
+		klog.Infof("driver responded but is not ready")
+		return
+	}
+	liveness.Set(1)
+	klog.Infof("Health check succeeded")
+}
+
+func recordLiveness(endpoint string, pollTime int) {
+	// register prometheus metrics
+	err := prometheus.Register(liveness)
+	if err != nil {
+		klog.Fatalln(err)
+	}
+
+	// get liveness periodically
+	go func() {
+		for {
+			getLiveness(endpoint)
+			// wait to poll metrics again
+			time.Sleep(time.Duration(pollTime) * time.Second)
+		}
+	}()
+}
+
+func Run(endpoint, port, livenessendpoint string, pollTime int) {
+	klog.Infof("Liveness Running")
+	// start liveness collection
+	recordLiveness(endpoint, pollTime)
+
+	// start up prometheus endpoint
+	addr := net.JoinHostPort("0.0.0.0", port)
```

Comment:
why this is running on `0.0.0.0` ? can we create a [service](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#communicating-with-daemon-pods) and make use of it?

Reply:
i have created a Service and ServiceMonitor that rook will deploy to plumb this all into Prometheus.